// @effect/vitest/dist/internal/internal.js
/**
 * @since 1.0.0
 */
import * as Cause from "effect/Cause";
import * as Duration from "effect/Duration";
import * as Effect from "effect/Effect";
import * as Exit from "effect/Exit";
import { flow, pipe } from "effect/Function";
import * as Layer from "effect/Layer";
import { isObject } from "effect/Predicate";
import * as Schedule from "effect/Schedule";
import * as Schema from "effect/Schema";
import * as Scope from "effect/Scope";
import * as fc from "effect/testing/FastCheck";
import * as TestClock from "effect/testing/TestClock";
import * as TestConsole from "effect/testing/TestConsole";
import * as V from "vitest";
const runPromise = /*#__PURE__*/Effect.fnUntraced(function* (effect, _ctx) {
  const exit = yield* Effect.exit(effect);
  if (Exit.isFailure(exit)) {
    const errors = Cause.prettyErrors(exit.cause);
    for (let i = 0; i < errors.length; i++) {
      yield* Effect.logError(errors[i]);
    }
  }
  return yield* exit;
}, (effect, _, ctx) => Effect.runPromise(effect, {
  signal: ctx?.signal
}));
/** @internal */
const runTest = ctx => effect => runPromise(effect, ctx);
const TestEnv = /*#__PURE__*/Layer.mergeAll(TestConsole.layer, /*#__PURE__*/TestClock.layer());
/** @internal */
export const addEqualityTesters = () => {
  V.expect.addEqualityTesters([]);
};
/** @internal */
const testOptions = timeout => typeof timeout === "number" ? {
  timeout
} : timeout ?? {};
/** @internal */
const makeTester = (mapEffect, it = V.it) => {
  const run = (ctx, args, self) => pipe(Effect.suspend(() => self(...args)), mapEffect, runTest(ctx));
  const f = (name, self, timeout) => it(name, testOptions(timeout), ctx => run(ctx, [ctx], self));
  const skip = (name, self, timeout) => it.skip(name, testOptions(timeout), ctx => run(ctx, [ctx], self));
  const skipIf = condition => (name, self, timeout) => it.skipIf(condition)(name, testOptions(timeout), ctx => run(ctx, [ctx], self));
  const runIf = condition => (name, self, timeout) => it.runIf(condition)(name, testOptions(timeout), ctx => run(ctx, [ctx], self));
  const only = (name, self, timeout) => it.only(name, testOptions(timeout), ctx => run(ctx, [ctx], self));
  const each = cases => (name, self, timeout) => it.for(cases)(name, testOptions(timeout), (args, ctx) => run(ctx, [args], self));
  const fails = (name, self, timeout) => V.it.fails(name, testOptions(timeout), ctx => run(ctx, [ctx], self));
  const prop = (name, arbitraries, self, timeout) => {
    if (Array.isArray(arbitraries)) {
      const arbs = arbitraries.map(arbitrary => {
        if (Schema.isSchema(arbitrary)) {
          return Schema.toArbitrary(arbitrary);
        }
        return arbitrary;
      });
      return it(name, testOptions(timeout), ctx =>
      // @ts-ignore
      fc.assert(
      // @ts-ignore
      fc.asyncProperty(...arbs, (...as) => run(ctx, [as, ctx], self)),
      // @ts-ignore
      isObject(timeout) ? timeout?.fastCheck : {}));
    }
    const arbs = fc.record(Object.keys(arbitraries).reduce(function (result, key) {
      const arb = arbitraries[key];
      if (Schema.isSchema(arb)) {
        result[key] = Schema.toArbitrary(arb);
      }
      result[key] = arb;
      return result;
    }, {}));
    return it(name, testOptions(timeout), ctx =>
    // @ts-ignore
    fc.assert(fc.asyncProperty(arbs, (...as) =>
    // @ts-ignore
    run(ctx, [as[0], ctx], self)),
    // @ts-ignore
    isObject(timeout) ? timeout?.fastCheck : {}));
  };
  return Object.assign(f, {
    skip,
    skipIf,
    runIf,
    only,
    each,
    fails,
    prop
  });
};
/** @internal */
export const prop = (name, arbitraries, self, timeout) => {
  if (Array.isArray(arbitraries)) {
    const arbs = arbitraries.map(arbitrary => {
      if (Schema.isSchema(arbitrary)) {
        throw new Error("Schemas are not supported yet");
      }
      return arbitrary;
    });
    return V.it(name, testOptions(timeout),
    // @ts-ignore
    ctx => fc.assert(fc.property(...arbs, (...as) => self(as, ctx)), isObject(timeout) ? timeout?.fastCheck : {}));
  }
  const arbs = fc.record(Object.keys(arbitraries).reduce(function (result, key) {
    const arb = arbitraries[key];
    if (Schema.isSchema(arb)) {
      throw new Error("Schemas are not supported yet");
    }
    result[key] = arb;
    return result;
  }, {}));
  return V.it(name, testOptions(timeout),
  // @ts-ignore
  ctx => fc.assert(fc.property(arbs, as => self(as, ctx)), isObject(timeout) ? timeout?.fastCheck : {}));
};
/** @internal */
export const layer = (layer_, options) => (...args) => {
  const excludeTestServices = options?.excludeTestServices ?? false;
  const withTestEnv = excludeTestServices ? layer_ : Layer.provideMerge(layer_, TestEnv);
  const memoMap = options?.memoMap ?? Effect.runSync(Layer.makeMemoMap);
  const scope = Effect.runSync(Scope.make());
  const contextEffect = Layer.buildWithMemoMap(withTestEnv, memoMap, scope).pipe(Effect.orDie, Effect.cached, Effect.runSync);
  const makeIt = it => Object.assign(it, {
    effect: makeTester(effect => Effect.flatMap(contextEffect, context => effect.pipe(Effect.scoped, Effect.provide(context))), it),
    prop,
    flakyTest,
    layer(nestedLayer, options) {
      return layer(Layer.provideMerge(nestedLayer, withTestEnv), {
        ...options,
        memoMap,
        excludeTestServices
      });
    }
  });
  if (args.length === 1) {
    V.beforeAll(() => runPromise(Effect.asVoid(contextEffect)), options?.timeout ? Duration.toMillis(Duration.fromInputUnsafe(options.timeout)) : undefined);
    V.afterAll(() => runPromise(Scope.close(scope, Exit.void)), options?.timeout ? Duration.toMillis(Duration.fromInputUnsafe(options.timeout)) : undefined);
    return args[0](makeIt(V.it));
  }
  return V.describe(args[0], () => {
    V.beforeAll(() => runPromise(Effect.asVoid(contextEffect)), options?.timeout ? Duration.toMillis(Duration.fromInputUnsafe(options.timeout)) : undefined);
    V.afterAll(() => runPromise(Scope.close(scope, Exit.void)), options?.timeout ? Duration.toMillis(Duration.fromInputUnsafe(options.timeout)) : undefined);
    return args[1](makeIt(V.it));
  });
};
/** @internal */
export const flakyTest = (self, timeout = Duration.seconds(30)) => pipe(self, Effect.scoped, Effect.sandbox, Effect.retry(pipe(Schedule.recurs(10), Schedule.while(_ => Effect.succeed(Duration.isLessThanOrEqualTo(Duration.fromInputUnsafe(_.elapsed), Duration.fromInputUnsafe(timeout)))))), Effect.orDie);
/** @internal */
export const makeMethods = it => Object.assign(it, {
  effect: makeTester(flow(Effect.scoped, Effect.provide(TestEnv)), it),
  live: makeTester(Effect.scoped, it),
  flakyTest,
  layer,
  prop
});
/** @internal */
export const {
  /** @internal */
  effect,
  /** @internal */
  live
} = /*#__PURE__*/makeMethods(V.it);
/** @internal */
export const describeWrapped = (name, f) => V.describe(name, it => f(makeMethods(it)));
